Add `use_newlines` and `newlines_max_depth`

gofastlily opened the pull request:
## Synopsis

Add the ability to enable/disable newlines in config.

Implement depth changes, the ability to set maximum depth in config, and `prnn()`/`prnnn()` from [this Dot Hop commit](https://github.com/russmatney/dothop/commit/1c6e31d24c6d62386f99cd696b89985540f8db12).


## Context

I really like this log library but prefer my logs to exist on a single line whenever possible.

In [discussion on this PR](https://github.com/russmatney/log.gd/pull/8#issuecomment-3233423233) Russ mentioned his changes from Dot Hop, and, given their relevance, are bundled in this PR.


## Changes

The following config value has been added to change the use of newlines project-wide, and defaults to `true`.

```gdscript
log_gd/config/use_newlines
log_gd/config/newline_max_depth
```

The following methods have been added to change the use of newlines in code.

```gdscript
Log.disable_newlines()
Log.enable_newlines()
Log.set_newline_max_depth(new_depth: int)
Log.reset_newline_max_depth()
```

Documentation has been updated to reflect the addition of `use_newlines` and `newline_max_depth` and their associated code.

Tests have been added to validate the use of newlines and the efficacy of the `use_newlines` and `newline_max_depth` config values.


## Other Changes

`Log.color_wrap()` now checks if `color_theme` is truthy before calling `color_theme.has_bg()`, which fixes many broken tests.  This change was needed to write unit tests for `use_newlines`-related logic.

`rebuild_config()` now keeps config values that have been set via code.  This matters because `rebuild_config()` would clear config values set before the first `to_printable()` call.  This was needed to allow setting config values in code before logging something for the first time.

Fixed a bug where `KEY_MAX_ARRAY_SIZE` was initialized as a `bool` instead of an `int`.  Not related but it was a quick fix while I was in the codebase.

## Author's Notes

I could split the two "other changes" out into their own pull requests if-needed, but I discovered the issues in the midst of adding `use_newlines` and fixed them in situ.

I'm also not sold on naming conventions so please let me know and I can update accordingly.

gofastlily committed:
Allow for single-line logs with Log.prn() and its aliases

russmatney commented:
Hi there! Thank you for the PR! Sorry for the state of the code, and thanks for the effort and fixes, they are much appreciated. I'd been getting increasingly anxious about having left things semi-stable (and I should be working on branches, not `main` directly).

I agree newlines tend to make these logs less readable - I've started toying with applying newlines at limited depths rather than everywhere, which I think works better. More to come on that soon (hopefully this weekend?) - I'd started a version of it in another repo: https://github.com/russmatney/dothop/commit/1c6e31d24c6d62386f99cd696b89985540f8db12

`use_newlines` as a configurable thing like this is great - I was actually surprised to see the `Log.warn/err/etc` aliases opt-in to newlines, and am thinking we should default those to single-line prints anyway.

What do you think of this: All the public print funcs default to single-lines, _except_ the family of `Log.prn`, `Log.prnn`, `Log.prnnn`, which add newlines at depths of 1, 2, and 3. Then `use_newlines` will impact `Log.warn/info/error/etc`, but not `Log.pr` itself?

Let me know what you think! I'll make time to dig into the code and likely merge this in the next day or two.

gofastlily commented:
Hi Russ!  Thanks for the quick reply!

Your suggestions sound good!  I'd like to try implementing them in my branch if that's okay.

I've got a quick lil' refactor I'd like to do on my changes as well, so I'll get a commit pushed up with my changes then a second with your suggestions soon.

russmatney commented:
> Hi Russ! Thanks for the quick reply!
> 
> Your suggestions sound good! I'd like to try implementing them in my branch if that's okay.
> 
> I've got a quick lil' refactor I'd like to do on my changes as well, so I'll get a commit pushed up with my changes then a second with your suggestions soon.

excellent, go for it!!

gofastlily committed:
Use Log.get_use_newlines() within Log.to_pretty()

gofastlily committed:
Implement prototype log features from Dot Hop

https://github.com/russmatney/dothop/commit/1c6e31d24c6d62386f99cd696b89985540f8db12

gofastlily committed:
Reorganize the example scene and ensure all showcase output gets printed

gofastlily commented:
2ecd6d3 is the refactor I was thinking about; moving the config check into `Log.to_pretty()`.

9639159 has the depth changes from Dot Hop, including tests and documentation updates.

92d14c2 is a bit of an opinionated refactor of `ExampleScene.gd` to ensure all output prints and all features are actually showcased.  The showcase output was well over the default for `network/limits/debugger/max_chars_per_second` and it made more sense to keep the default there instead of changing project defaults in a demo.

russmatney commented:
🔥 Huge! Thanks for the work! i've skimmed it and it looks good overall. Let me spend a bit more time with it tonight/tomorrow and I'll get it merged soon after.

One thing I can hit afterwards - i suspect (but may be wrong) that the delimiter index and depth vals are the same thing - i was intending to drop delimiter-index in favor of using the depth directly for the color cycles. Happy to poke at it myself, or feel free if you want to and have the context already

gofastlily committed:
Remove delimiter_index in favor of newline_depth

gofastlily commented:
I had a few minutes so  I figured I'd replace `delimiter_index` in favor of `newline_depth` and removed the old references to `delimiter_index`.

All the tests that were passing still pass.

russmatney commented:
@gofastlily You rule! much appreciated!

I'll ping you if I run into anything while digging into all this in more detail - and let me know if you have any thoughts/feature ideas/bugs or what not!

gofastlily commented:
Thank you so much!!

I've been considering a setting for log levels that only prints the level selected and below, eg. if the level is set to WARN, `Log.info()` won't print anything.

I'll probably tinker with that this evening if it sounds interesting to you.

russmatney commented:
sure, feel free - that'd be a great feature to add.

My dream is toggling things like that directly with output UI buttons in the godot editor - and even extending that to toggle per module or file - but i've yet to dig into what hooking into that looks like (or if it's accessible from plugin space). It's likely a can of worms, no worries either way, just something I'm curious to learn more about).

russmatney committed:
fix: ensure pretty=false on warning/error messages

We want to skip-pretty on the push_warning/push_error calls to keep them
readable in the debugger view.

russmatney commented on addons/log/log.gd:
```diff
@@ -502,52 +564,53 @@ static func log(msg: Variant, msg2: Variant = "ZZZDEF", msg3: Variant = "ZZZDEF"
-	print_rich(Log.to_printable(rich_msgs, {stack=get_stack(), newlines=true}))
-	var m: String = Log.to_printable(msgs, {stack=get_stack(), newlines=true, pretty=false})
+	print_rich(Log.to_printable(rich_msgs, {stack=get_stack()}))
+	var m: String = Log.to_printable(msgs, {stack=get_stack(), pretty=true})
```
I pushed a commit flipping this back to false for warns/errors, but i realized I should ask was there a reason to set this to true? On my end it makes reading the errors/warnings a mess 

<img width="1299" height="155" alt="Screenshot 2025-08-28 at 7 04 39 PM" src="https://github.com/user-attachments/assets/feb2b3e1-92bb-4ddb-beda-0fa3ea4d5441" />

vs

<img width="624" height="158" alt="Screenshot 2025-08-28 at 7 08 26 PM" src="https://github.com/user-attachments/assets/564b4d97-fdbe-40a6-aa9b-69af373a22ff" />

gofastlily replied:
That was just a goof on my part, good catch!

russmatney committed:
fix: drop/simplify some tests, drop cold-fire theme default

Gets tests passing again! heh. Will return to these in a future branch.

The coldfire theme is cool, but not quite ready yet.

russmatney committed:
fix: just disable_colors on warns/errors

russmatney committed:
docs: add contributors section

russmatney commented:
Caught one thing around the warnings/errors - they should `disable_colors` to stay readable in the debugger.

Thanks a ton for the example scene rewrite and for pulling in those `prnnnn`s from that Dot Hop commit! I'm really liking the `prn` vs `prnn` vs `prnnn` bit :fire:

<img width="361" height="77" alt="Screenshot 2025-08-28 at 7 10 19 PM" src="https://github.com/user-attachments/assets/f0484343-2ad6-418b-92e9-59ce5c31c6fc" />
<img width="848" height="423" alt="Screenshot 2025-08-28 at 7 10 13 PM" src="https://github.com/user-attachments/assets/85a24eff-b009-4a97-86de-257a8bd7468b" />

Also thank you for the doc updates in all the places!

I simplified and/or dropped a handful of tests to get them passing again, and I'm pledging to work on branches and keep `main` more stable. Thanks for the inspiration!

The testing strategy has been tricky to maintain, and is now outdated since the move to resource based color themes - I'll have to think through a better strategy - comparing the printed strings in the test output is super annoying. I suspect smaller, wiser unit tests will give better coverage for less.

The config / rebuild_config / etc flow is kind of a mess... I'm open to better ways to do all of that, should probably look into how other addons handle it.

Thanks a ton for this PR! Feel free to open more if you have thoughts/opinions.

I'll make time to clean a few things up and cut a new release this weekend.
